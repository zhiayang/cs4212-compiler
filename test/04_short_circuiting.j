// 04_short_circuiting.j

class Main
{
	Void main()
	{
		println(new Foo().test(10, 420));
	}
}

class Foo
{
	String test(Int x, Int y)
	{
		if(effect1() || effect2())  { println("uwu"); }
		else                        { println("??????????"); }

		while((x >= 0 && x < 7) || effect3(x)) { println("owo"); x = x - 1; }


		if(effect2() && effect3(69)) { println("oh no"); }
		else                         { println("asdf"); }

		if(!effect2() && effect3(4)) { println("oh no"); }
		else                         { println("bsdf"); }


		// all must be tried
		if(effect2() || effect2() || effect2() || effect2() || effect2() || effect1()) { println("ok"); }
		else                                                                           { println("oh no"); }

		if(effect1() && effect1() && effect1() && effect1() && effect1() && effect1()) { println("ok"); }
		else                                                                           { println("oh no"); }

		return "kekw";
	}

	Bool effect1()
	{
		println("effect 1");
		return true;
	}

	Bool effect2()
	{
		println("effect 2");
		return false;
	}

	Bool effect3(Int k)
	{
		println("effect 3");
		return k > 5;
	}
}
