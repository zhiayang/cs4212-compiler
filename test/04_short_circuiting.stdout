effect 1
uwu
effect 3
owo
effect 3
owo
effect 3
owo
effect 3
owo
owo
owo
owo
owo
owo
owo
owo
effect 3
effect 2
asdf
effect 2
effect 3
bsdf
effect 2
effect 2
effect 2
effect 2
effect 2
effect 1
ok
effect 1
effect 1
effect 1
effect 1
effect 1
effect 1
ok
kekw
